src/js/actions/sampler.js

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2014 Adobe Systems Incorporated. All rights reserved.
+ *  
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"), 
+ * to deal in the Software without restriction, including without limitation 
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, 
+ * and/or sell copies of the Software, and to permit persons to whom the 
+ * Software is furnished to do so, subject to the following conditions:
+ *  
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+define(function (require, exports) {
+    "use strict";
+
+    var Promise = require("bluebird"),
+        Immutable = require("immutable");
+
+    var uiUtil = require("js/util/ui"),
+        locks = require("js/locks"),
+        layerFXActions = require("./layereffects"),
+        typeActions = require("./type"),
+        shapeActions = require("./shapes");
+
+    /**
+     * Process a single click from the sampler tool.
+     * Based on the clicked leaf layer type, will sample the property from it
+     * 
+     * @private
+     * @param {Document} doc Document model
+     * @param {number} x Offset from the left window edge
+     * @param {number} y Offset from the top window edge
+     * @return {Promise}
+     */
+    var click = function (doc, x, y, secondary) {
+        var uiStore = this.flux.store("ui"),
+            coords = uiStore.transformWindowToCanvas(x, y),
+            layerTree = doc.layers;
+        
+        return uiUtil.hitTestLayers(doc.id, coords.x, coords.y)
+            .bind(this)
+            .then(function (hitLayerIDs) {
+                var hitLayerMap = hitLayerIDs.reduce(function (layerMap, id) {
+                        layerMap[id] = true;
+                        return layerMap;
+                    }, {}),
```

Comment:
I think we generally prefer using `Map` or, in this case, `Set` for this sort of thing instead of objects.